pkg/cmd/pulumi/new.go

```diff
@@ -457,6 +468,55 @@ func runNew(ctx context.Context, args newArgs) error {
 	return nil
 }
 
+// Retrieve a Copilot template from the given URL **including an auth token for Pulumi Cloud**
+// We only want to send this token to the app.pulumi.com domain.
+func retrieveCopilotTemplate(
+	ctx context.Context,
+	unsafeTemplateURL string,
+	opts display.Options,
+) (workspace.TemplateRepository, error) {
+	// Get a token for the copilot template cloud (only https://api.pulumi.com is supported for now).
+	lm := httpstate.NewLoginManager()
+	account, err := lm.Login(ctx, "https://api.pulumi.com", false /* insecure */, "pulumi",
+		"Copilot templates", httpstate.WelcomeUser, false /* setCurrent */, opts)
+	if err != nil {
+		return workspace.TemplateRepository{}, fmt.Errorf("logging in: %w", err)
+	}
+
+	u, err := url.Parse(unsafeTemplateURL)
+	if err != nil {
+		return workspace.TemplateRepository{}, fmt.Errorf("parsing template URL: %w", err)
+	}
+
+	templateURL := url.URL{
+		Scheme: "https",
+		Host:   "app.pulumi.com",
+		Path:   u.Path,
+	}
+
+	return workspace.RetrieveZIPTemplates(templateURL.String(), func(req *http.Request) {
+		req.Header.Set("Authorization", "token "+account.AccessToken)
+	})
+}
+
+// Test whether we should try and retrieve a template from a Copilot URL.
+// They require an auth token and so are handled a bit differently they look like this:
+// https://app.pulumi.com/ai/api/bundle/{org}/{conversation}/{program}.zip
+func isCopilotTemplateURL(templateNamePathOrURL string) bool {
```

Comment:
Rather than have this check, can we have `api.pulumi.com` proxy this request and select the appropriate HTTP backend to send requests to Pulumi Copilot?

This does add an additional hop, but I think keeps us aligned with self-hosted backend support. 

This would mean that we need a URL beginning with `https://api.pulumi.com/...` to handle the request, but we have done that before with `pulumi new --ai`.

Here in the service backend implementation (also called "httpstate"), we define the path to use:

https://github.com/pulumi/pulumi/blob/7c14d59c2bd1392a1fda9110ec3f6b01a3d0fea5/pkg/backend/httpstate/client/client.go#L197-L200

And later, we define a method for requesting an AI template:

https://github.com/pulumi/pulumi/blob/7c14d59c2bd1392a1fda9110ec3f6b01a3d0fea5/pkg/backend/httpstate/client/client.go#L1279-L1296

I can point you to the backend code in Slack for this

Reply:
Sorry - one more layer of indirection, that's surfaced as a method on the backend interface:

https://github.com/pulumi/pulumi/blob/7c14d59c2bd1392a1fda9110ec3f6b01a3d0fea5/pkg/backend/httpstate/backend.go#L1193C24-L1204

Which is called here:

https://github.com/pulumi/pulumi/blob/7c14d59c2bd1392a1fda9110ec3f6b01a3d0fea5/pkg/cmd/pulumi/new_ai.go#L135-L138

One quirk of this which I think we'll want to work out though is that early on in new.go we select the backend _before_ looking at the template URL:

https://github.com/pulumi/pulumi/blob/7c14d59c2bd1392a1fda9110ec3f6b01a3d0fea5/pkg/cmd/pulumi/new.go#L165-L189

https://github.com/pulumi/pulumi/blob/7c14d59c2bd1392a1fda9110ec3f6b01a3d0fea5/pkg/cmd/pulumi/util.go#L166-L174

But this is an interesting question, I'm not sure how I'd solve it:

A template could be for a public URL (like `https://github.com/...`) or an AI URL (`https://api.pulumi-self-hosted.foo.corp/...`)...